### PR TITLE
DQMStore preparations for concurrent booking - phase 2

### PR DESCRIPTION
**Pull request by knoepfel**
This PR includes many changes to the `DQMStore`, built off of PR #22614 .  I have made the following changes:

- `DQMStore` no longer has global `run_` and `moduleId_` data members.  The run and module-ID numbers are now owned by the `IBooker` transactional object.
- Any per-run/module histograms are booked via the `IBooker`.   To retain backwards compatibility, **I have made the assumption that all direct uses of the `DQMStore::book*` functions are intended to book global histograms.  If this assumption is incorrect, please show me the counterexamples.**
- The container for storing `MonitorElement`s is no longer an `std::set` but an `std::map`.  To allow for concurrent booking, this will need to change to a concurrent unordered map.  That will be the next stage.
- The duplication of `std::string const&/char const*` booking-function overloads has been replaced by a single set of functions that takes a `char_string const&` argument.  The `char_string` class has implicit constructors allowing the conversions from `char const*`, `std::string`, and `TString` types.  This cuts down drastically on the overload set.  Ideally, in C++17, we could use `std::string_view` instead--but that may be a change for another time.

With all of the above changes, and given the assumption I make about raw usage of `DQMStore::book*`, I believe there are no breaking changes.  At least, none of my local builds (including many dependent packages) have exposed any.

The next phase for concurrent booking will involve removing the global `DQMStore::pwd_` data member and introducing a proxy class to ensure synchronization of the directory between `IBooker` and `IGetter` instances.

### CODE REVIEW

**Comment by schneiml** [comment]
Hi Kyle, thanks for your effort to clean up the `DQMStore`!

I have a few concerns that we found while discussing this migration:
- The usual: Keep in mind that DQMGUI uses the DQMNet/DQMStore code as well. It uses a pinned version, so as long as the file formats remain compatible, everything is fine. But having the verisons diverge is not a long term solution. There are efforts to remove this code-level dependency (@rovere is the expert).
- Be aware of the subtle differences between "legacy" and "threaded" mode. Figuring out these is an ongoing research effort, though I think we have a fairly good understanding by now.
- That said, your assumption that everything using the "legacy" booking interface books global MEs is correct (as far as I know).
- But, be aware of the arcane details in the query logic! These are why the `enableMultiThread` flag is still there, even though it does things that have nothing to do with what the name suggests. I made an effort to remove that and de-duplicate the query logic [1], but these are far diverged by now. We abandoned this due to the risk of breaking untestable tools for datataking. For 10_2, there are some chances that we can touch this, though some people would sure prefer that to happen no earlier than the long shutdown.

On to more concrete suggestions:

- I like the booking interface code changes.
- Moving the booking state into IBooker is also a good thing.
- The style changes are (hopefully) harmless, but make reading the diff quite hard.  Therefore, I'd like to see that as an independent PR that should not cause any trouble.
- The changes towards `std::map` and to the Key objects seem somewhat dangerous. _Probably_ they are all fine, but that is hard to tell with certainity. This is another reason why I'd like to see this PR split up, tough I see the order of the commits makes that nontrivial...
- For the future, there is the idea fluctuating around to actually split the container with the MEs by module, rather than keeping the module ID in the key. That would solve most concurrecny issues without fancy container types, I think. Also we could retain most of the legacy interface for legacy modules, with only a slight change in semantics, that most of the remaining legacy code (comissioning tools etc.) won't notice.
- This "per-module DQMStore" could also nicely integrate with edm dependency handling. But the details on why this works are fairly non obvious at the moment. I might write down what we planned there at some point.

[1] https://github.com/schneiml/cmssw/commits/dqm-token-migration

tl;dr: We want most of these changes, but I'd feel saver if you could cut out the "functional" changes into a second PR.

**Comment by knoepfel** [comment]
@schneiml Thanks, Marcel, for your comments.  I agree that as far as the style changes are concerned, etc., those could have waited--my apologies for that.  My editor also removes extraneous whitespace and replaces tabs with whitespaces, which is why (e.g.) `DQMNet` seemed to had such a significant diff, even though the only functional change was to add the `KeyObject` class.

I think I may be able to reorder commits, separating cleanups from functional changes, eventually resulting in two PRs along the lines you suggest.  If you'd prefer to reject this PR, that's fine with me.  I appreciate your guidance.

**Comment by fabiocos** [comment]
@knoepfel @schneiml what is the plan for this PR?

**Comment by knoepfel** [comment]
My plan was to change this PR to be a cleanup-only PR with not functional changes.  My other work commitments have gotten in the way of this, however.  Practically speaking, I don't see this PR getting updated for a couple weeks yet.

**Comment by fabiocos** [comment]
@knoepfel @schneiml is there any news about this PR?

**Comment by schneiml** [comment]
@fabiocos @knoepfel I picked up work on the DQM framework again, currently aiming for concurrent lumisections (but this could imply larger refactoring as well). I guess this specific PR could be closed, and a new one opened once there is new progress.

**Comment by fabiocos** [comment]
@schneiml @knoepfel ok, thanks, I close this PR waiting for further developments
